### ISSUE TITLE
Ocentric (XYZ) Lidar Data Do No Render Properly

### ISSUE THREAD

**Issue by jlaura** [qgis/QGIS#57659]
### What is the bug or the crash?

COPC formatted lidar data, with proper projection WKT projection information do not render properly. Sample data are available either from [https://na-c.entwine.io/lola.copc.laz](https://na-c.entwine.io/lola.copc.laz) or our LOLA in XYZ [https://asc-moon.s3-us-west-2.amazonaws.com/hobu/LolaRDR_0N15N_0E15E.copc.laz](https://asc-moon.s3-us-west-2.amazonaws.com/hobu/LolaRDR_0N15N_0E15E.copc.laz). These are a global and 15˚ tile of some lunar orbiter laser altimeter data. The WKT projection defined properly in the data is:

```GEODCRS["IAU_2015MoonXYZ",
    DATUM["Moon(2015)-Sphere",
        ELLIPSOID["Moon(2015)-Sphere",1737400,0,
            LENGTHUNIT["metre",1]]],
    PRIMEM["Reference meridian",0,
        ANGLEUNIT["degree",0.0174532925199433]],
    CS[Cartesian,3],
        AXIS["(X)",geocentricX,
            ORDER[1],
            LENGTHUNIT["metre",1]],
        AXIS["(Y)",geocentricY,
            ORDER[2],
            LENGTHUNIT["metre",1]],
        AXIS["(Z)",geocentricZ,
            ORDER[3],
            LENGTHUNIT["metre",1]]]```

Depending on the CRS set for the project, incorrect rendering examples are (project SRS set to the above WKT string):
![Screenshot 2024-06-03 at 11 02 18 AM](https://github.com/qgis/QGIS/assets/2105468/713979d5-f7d1-4c6a-a911-e54e105028af)

and (project SRS set to IAU_2015:30100):

![Screenshot 2024-06-04 at 10 31 42 AM](https://github.com/qgis/QGIS/assets/2105468/3af334ad-efac-402d-a960-1ef69c359ed1)


### Steps to reproduce the issue

Add one of the above linked COPC files to QGIS as a point cloud source. The data are freely available and will stream efficiently to the installation.

### Versions

3.34.x, 3.36.x

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

See [moonlidar.com](https://moonlidar.com) for the correct rendering of these data.

I can only surmise that the issue is some combination of body-centered body-fixed (XYZ meters) rendering and/or the fact that the lunar sphere is used with a 1737.4 km radius while the QGIS code base makes some Earth radii assumption.

**Comment by uclaros** [comment]
@jlaura Can you please describe what is wrong with the rendering? The link you provided does not have a 2D view of the data to compare. 
Furthermore, it uses _Altitude_ to style the points which is not an attribute in your point cloud files.

**Comment by jlaura** [comment]
@uclaros Absolutely! [Here](https://stac.astrogeology.usgs.gov/browser-dev/#/api/collections/lunar_orbiter_laser_altimeter/items/LolaRDR_-45N-30N_330E345E?.asset=asset-thumbnail) is a link to a thumbnail that is generated from these data if they are encoded in the COPC file as lat/lon/height above the datum (which stores poorly), as opposed to XYZ encoding that they are using now. Yes, this thumbnail has been gridded from the point data and is offered only as an example of what these data should render as. I have also included that image inline.

<img width="379" alt="Screenshot 2024-06-06 at 5 56 14 AM" src="https://github.com/qgis/QGIS/assets/2105468/331cd399-6dc8-441f-9d3b-88a3af2b7b22">

[moonlidar.com](https://moonlidar.com) shows the correct point rendering from the first COPC file linked in the original issue. One sees the lunar topography and not the general red-to-blue, left-to-right gradient that QGIS renders.

As for styling, the points should be x,y and then stylized with the z dimension. If these were to be converted using something like pyproj, I would use:

```python
bcbf = pyproj.Proj(proj='geocent'j, a=1737400, b=1737400)
lla = pyproj.Proj(proj='lonlat', a=1737400, b=1737400)
trans = pyproj.Transformer.from_proj(bcbf, lla)
lon, lat, elev = trans.transform(x,y,z)
```

where that code snippet assumes the data come in in geocentric XYZ and the desired coordinates for rendering are lat, lon, elevation.

Thanks for your interested in the issue! Please let me know if anything is unclear, I have misunderstood anything, or if I can provide more information.

**Comment by wonder-sk** [comment]
@jlaura with the upcoming QGIS 3.44 it is possible to use globe 3D view and stream https://na-c.entwine.io/lola.copc.laz to QGIS 🎉   (there are some rough edges, but it works somehow)

![Image](https://github.com/user-attachments/assets/8807fbb4-7cee-4386-8dc2-e577655ef0f9)

**Comment by jlaura** [comment]
@wonder-sk Woohoo! 🚀 That is absolutely terrific! That seem is 100% in the data (an artifact). The rendering looks terrific.

I will make sure to widely publicize. 

Again, terrific!! 🚀 🥇 🎉.

**Comment by jlaura** [comment]
I was recently testing this in QGIS 4.0 and uncovered some interesting behavior. As long as the project CRS matches the X,Y,Z definition, QGIS renders as expected (again exciting!). 

<img width="1305" height="1251" alt="Image" src="https://github.com/user-attachments/assets/ed9849ce-d03e-4d2d-8a6e-508498b26c23" />

But, if I add a WMS and adjust the CRS to be something standard (e.g., IAU_2015:30100), the XYZ data are "destroyed". 

<img width="1610" height="1396" alt="Image" src="https://github.com/user-attachments/assets/17da7e72-684e-436c-b25c-f9189559efd8" />

I suspect this is an on-the-fly projection issue (?) and that it might relate to the fact that, while 30100 is well defined inside PROJ, the point cloud definition (below) is not. 

  GEODCRS["IAU_2015MoonXYZ",DATUM["Moon(2015)-Sphere",ELLIPSOID["Moon(2015)-Sphere",1737400,0,LENGTHUNIT["metre",1,ID["EPSG",9001]]]],CS[Cartesian,3],AXIS["(X)",geocen
  tricX,ORDER[1],LENGTHUNIT["metre",1]],AXIS["(Y)",geocentricY,ORDER[2],LENGTHUNIT["metre",1]],AXIS["(Z)",geocentricZ,ORDER[3],LENGTHUNIT["metre",1]],ID["IAU",30000,20
  15]]

Perhaps someone with my knowledge of the internals of projection on the fly and SRS handling can provide insight?

**Comment by JanCaha** [comment]
@jlaura What exactly do you mean by this:

> I was recently testing this in QGIS 4.0 and uncovered some interesting behavior. As long as the project CRS matches the X,Y,Z definition, QGIS renders as expected (again exciting!).

I loaded [ https://asc-moon.s3-us-west-2.amazonaws.com/hobu/LolaRDR_0N15N_0E15E.copc.laz]( https://asc-moon.s3-us-west-2.amazonaws.com/hobu/LolaRDR_0N15N_0E15E.copc.laz) the dataset and it renderes like in you first post (badly rendered). What do you need to set project CRS to to get the correct render?

**Comment by JanCaha** [comment]
@jlaura could share the data used in the last screenshot? That seems to contain some error that we could replicate to trace it back.

**Comment by jlaura** [comment]
Absolutely! I have a smaller LAZ file with the same projection information (BCBF). I incorrectly assumed that, because the linked file was working, the global file would as well. 

You should be able to pull the LAZ file from [here](https://asc-mars.s3.us-west-2.amazonaws.com/QGIS/57659/TC1W2B0_01_00773S607E0844__TC2W2B0_01_00773S616E0843.copc.laz). It's a ~45MB COPC file (hosted on S3).

**Comment by uclaros** [comment]
@jlaura is the wms server also public? Can you share the url for that too?

**Comment by jlaura** [comment]
@uclaros  Sure thing! It's the USGS planetaryMaps WMS. Here is the direct link to the equipment moon service: https://planetarymaps.usgs.gov/cgi-bin/mapserv?map=/maps/earth/moon_simp_cyl.map&service=WMS&request=GetCapabilities

**Comment by JanCaha** [comment]
@jlaura The issue here is that the WMS service says that it can work with WGS84, which is used as default here. That causes issues because even if you set projects CRS IAU_2015:30100 the WMS cannot be transformed to it (because the source is EPSG:4326 and there is no conversion from that to Moon IAU:30100). That leads to WMS being used AS-IS without transform, which does not make sense, as it does not fit you other layer.

That being said, the correct way to use the WMS would be to use **Data Source Manager** (from **Layer** menu).  There you can load the layer with correct CRS (see right bottom of the image) if the WMS service and layer actually supports it.
<img width="1000" height="810" alt="Image" src="https://github.com/user-attachments/assets/dc2062ba-669d-4d14-bc89-3f4efd63027d" />

That brings up the last issue. Not many WMS servers actually correctly support this as this moment. If you change the CRS of the WMS that you provided to IAU_2015:30100 you actually do not get any image (it bland or white) but the issue is in the WMS server.

We recently propose change in this PR #65586 that fixes it for QGIS WMS Server. So there should be at least one WMS server available to serve non-Earth data correctly.

Anyways the COPC layer loads correctly on it own and looks ok. There few more pending changes #65086 and #65026 that when merged into QGIS will make these layer work even better.

**Comment by JanCaha** [comment]
@jlaura What i mentioned above stands, but I actually noticed something strange with the data.

I selected one random point from the COPC datase.
```
x=82554.52615704626, y=835944.3699126304, z=-550.866130856155
```
But then i convert it from the WKT coordinate system specified in COPC file to IAU_2015:30100 I got
```
x=84.36, y=-0.04, z=-897388.96
```

Which is like halfway inside the Moon Ellipsoid looking at the Z coordinate. That could be source of some of those rendering issues that you are seeing. Can you verify that the CRS there is actually correct and properly set and that the X,Y, Z values of points are correct?
